_posts/2016-06-17-using-xhprof.md

```diff
@@ -0,0 +1,66 @@
+
+### Take a step back and plan your test scenarios
+
+Before digging in to optimizing your module code, you need to take a step back and think about the big picture. First, what are you optimizing for? Many optimizations involve a tradeoff between time and memory usage. Are you trying to reduce overall run-time at the expense of more memory? Is keeping the memory footprint of a given function down more important? In order to answer these questions you need to think about the overall context in which your code is running. In my case, I was optimizing a background import module which was run via cron, so the top priority was that the memory usage and number of database optimizations were low enough not to impact the user-facing site performance.
+
+Second, what use case for your code are you profiling? If this is a single method call, what arguments will be passed? If you're profiling page loads on a website, which pages are you profiling? In order to successfully track whether the changes you're making are having an impact on the metrics you're concerned about, you need to be able to narrow down the possible use cases for your code into a handful of most-likely real world scenarios which you'll actually choose to track via the profiler.
+
+Now it's time to get organized. Write a simple test script so that you can easily run through all your use cases in an automated way -- this is not strictly necessary, but it will save you a lot of work and potential error as you move through the process. In my case, I was testing a drush command hook, so I just wrote a bash shell script which executed the command three times in each of two different ways. For profiling page loads, I would recommend using [Apache JMeter](https://jmeter.apache.org/) - and you'll need to consider whether you want to force an uncached page load by passing a random dummy query parameter. Ideally, you should be running each scenario a few times so that you can then average the results to account for any small variations in run-time.
+
+### Keep things organized
+
+Keeping your different runs organized is probably the most important part of successfully profiling module code using XHProf! Each run has a unique run ID, but **you** are solely responsible for knowing which use case scenario and which version of the codebase that run ID corresponds to. I set up a basic spreadsheet in OpenOffice where I could paste in run numbers and basic run stats to compare (but there's almost certainly a nicer automated way to do this than what I used).
+
+<img src="/assets/img/blog/xhprof-results-spreadsheet.png" alt="Screenshot of an OpenOffice spreadsheet summarizing XHProf results for various runs">
```

Comment:
This image does have some client initials in it. I'm not sure if that matters, although the rest of the post remains fairly client-agnostic. It probably doesn't matter, just pointing it out ;)